src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs

```diff
@@ -1,93 +1,213 @@
+            => new StatementRange(ifNode.Statement, ifNode.Statement);
+
+        protected override bool IsStatementContainer(SyntaxNode node)
+            => node.IsKind(SyntaxKind.Block, SyntaxKind.SwitchSection);
+
+        protected override bool IsNoOpSyntaxNode(SyntaxNode node)
+            => node.IsKind(SyntaxKind.Block, SyntaxKind.EmptyStatement);
+
+        protected override bool IsStatement(SyntaxNode node)
+            => node is StatementSyntax;
+
+        protected override SyntaxNode GetNextExecutableStatement(SyntaxNode node)
+            => CSharpSyntaxFactsService.Instance.GetNextExecutableStatement(node);
+
+        protected override StatementSyntax GetIfBody(IfStatementSyntax ifNode)
+            => ifNode.Statement;
+
+        protected override StatementSyntax GetEmptyEmbeddedStatement()
+            => SyntaxFactory.Block();
+
+        protected override StatementSyntax GetElseBody(IfStatementSyntax ifNode)
+            => ifNode.Else.Statement;
+
+        protected override TextSpan GetHeaderSpan(IfStatementSyntax ifNode)
+        {
+            return TextSpan.FromBounds(
+                ifNode.IfKeyword.SpanStart,
+                ifNode.CloseParenToken.Span.End);
+        }
+
+        protected override bool CanControlFlowOut(SyntaxNode node)
         {
-            var ifStatement = token.GetAncestor<IfStatementSyntax>();
-            if (ifStatement == null || ifStatement.Else == null)
+            switch (node.Kind())
             {
-                return null;
+                case SyntaxKind.SwitchSection:
+                case SyntaxKind.LocalFunctionStatement:
+                case SyntaxKind.SetAccessorDeclaration:
+                case SyntaxKind.GetAccessorDeclaration:
+                case SyntaxKind.AddAccessorDeclaration:
+                case SyntaxKind.RemoveAccessorDeclaration:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.ConstructorDeclaration:
+                case SyntaxKind.DestructorDeclaration:
+                case SyntaxKind.OperatorDeclaration:
+                case SyntaxKind.ConversionOperatorDeclaration:
+                case SyntaxKind.AnonymousMethodExpression:
+                case SyntaxKind.SimpleLambdaExpression:
+                case SyntaxKind.ParenthesizedLambdaExpression:
+                case SyntaxKind.DoStatement:
+                case SyntaxKind.WhileStatement:
+                case SyntaxKind.ForStatement:
+                case SyntaxKind.ForEachStatement:
+                case SyntaxKind.ForEachVariableStatement:
+                    return false;
             }
 
-            var span = TextSpan.FromBounds(ifStatement.IfKeyword.Span.Start, ifStatement.CloseParenToken.Span.End);
-            if (!span.IntersectsWith(textSpan))
+            return true;
+        }
+
+        protected override SyntaxList<SyntaxNode> GetStatements(SyntaxNode node)
+        {
+            switch (node)
```

Comment:
these seems like generic enough to be in ISyntaxFactService?

Reply:
I can do that.

(pending)

Reply:
Please leave here.  SyntaxFacts is not for generic stuff.  It's for providing a unified view of hte *existing* VB and C# Syntax APIs on top of any type of node.  There is no GetStatements helper in C# or VB that operates over blocks or switches.  So this is code specific to this feature.